### DISCUSSION TITLE
Object Deletion and Saving?

### DISCUSSION THREAD

**Discussion by Validus-Tommy** [kibook/spooner#32] [Q&A]
As the title suggest, how would one go about saving/streaming object deletion/cleanup? I currently want to clean up and delete objects in a burnt out area, I can delete them in my test server but as soon as I walk a few feet the deleted objects return, do I need to do singular areas at a time?

**Comment by kibook** [comment]
There are different ways you can remove objects permanently, and I would approach the solutions in this order:

1. Disable the ymap (IPL) or interior entity set that contains the objects.
2. Remove the objects from the ymap or ytyd and stream the modified ymap or ytyd to replace the original.
3. Delete the objects via a script.

1 being the best solution, 3 being the "worst", in terms of efficiency at least.

If the objects you want to remove are all part of one ymap or interior entity set, simply disabling that ymap or interior entity set is likely the best solution. If the ymap or interior entity set contains additional things you don't want to get rid of, then editing the ymap or ytyd files to remove only the objects you want to remove would be better. If you're not able to do either of those, then it is possible to create a script which deletes the objects any time they spawn to effectively remove them permanently. This last solution is not very efficient because it means you have some code running in the background continuously, detecting and deleting objects, playing cat and mouse with the game's spawning system.

Spooner doesn't have any kind of features to help with the first two solutions, they require changing base game files. It can however indirectly help with the third option. I have another resource for RedM called [objectloader](https://github.com/kibook/redm-objectloader), which is meant to load Lambdarevolution map editor maps on a RedM server. The Lambdarevolution map format, unlike ymap, has the ability to specify objects that should be removed, not just new objects that should be added. When objectloader sees one of these `<DeletedObject>` elements in the map editor XML, it will attempt to find an object of the specified type at the specified coordinates, and if it finds it, it will delete it. It does this repeatedly so that any time the object gets respawned by the game, it is deleted again.

The "Map Editor XML" export format in spooner allows you to export your database to the format used by objectloader. To include *deleted* objects in the export, there is one additional step. You must check the "Save/load deletions" option in the saved DBs menu (J) *before* you delete the objects, in order for spooner to save a list of what objects you delete. Then when you export to Map Editor XML format, you should see some `<DeletedObject>` elements in the output.

If you can be more specific about what objects you're trying to delete and where the "burnt out area"  is, I or someone else might be able to look further into this and see which of these solutions will work. For example, we could find the ymap you need to disable if the things you want to delete are all in one.
